### PR TITLE
On PRs from external users, dismiss stale reviews when a new commit is pushed.

### PR DESCRIPTION
**Pull request by jonsimantov**
With our new project settings, when a PR is approved, that approval will stay valid even if further commits are pushed into that PR's branch.

For external users, though, we want to invalidate those approvals if any additional commits are made subsequent to the original approval. In this case, we consider anyone without "admin" access to be an "external user", and any PR from a forked repo to be an "external PR". In either of those scenarios, stale approvals will be dismissed.

This PR adds a workflow that runs on pull_request_target, which runs inside the `main` branch and has enough access to dismiss reviews. It also adds a script that allows you to dismiss reviews on a PR.

### CODE REVIEW

File: scripts/gha/dismiss_reviews.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  logging.debug("Found %d reviews", len(reviews))
+
+  # Filter out already-dismissed reviews.
+  reviews = [r for r in reviews if r['state'] != 'DISMISSED']
+  # Filter by review_state if specified.
+  if FLAGS.review_state != 'ANY':
+    reviews = [r for r in reviews if r['state'] == FLAGS.review_state]
+  # Filter by reviewer's username, if specified.
+  if FLAGS.reviewer:
+    reviews = [r for r in reviews if r.user.login == FLAGS.reviewer]
```
Comment (jonsimantov) [inline scripts/gha/dismiss_reviews.py:72]: Whoops, this is a bug - will change to r['user']['login']

File: scripts/gha/dismiss_reviews.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A utility to dismiss PR reviews.
+
+USAGE:
```
Comment (sunmou99) [inline scripts/gha/dismiss_reviews.py:17]: nit: update usage

File: .github/workflows/checks_secure.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+on:
+  pull_request_target:
+    types: [synchronize]
+
+jobs:
+  dismiss_stale_approvals:
+    # Dismiss stale approvals for non-admins or if this PR comes from a fork.
+    runs-on: ubuntu-latest
+    # Only if another commit was added to the PR.
+    if: github.event.action == 'synchronize'
```
Comment (sunmou99) [inline .github/workflows/checks_secure.yml:13]: Is it necessary? Isn't "synchronize" the only trigger?

File: .github/workflows/checks_secure.yml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    steps:
+      - name: Check user permission
```
Comment (sunmou99) [inline .github/workflows/checks_secure.yml:15]: output?

File: scripts/gha/dismiss_reviews.py (hunk 2 above)
Comment (jonsimantov) [reply scripts/gha/dismiss_reviews.py:17]: Done.

File: .github/workflows/checks_secure.yml (hunk 3 above)
Comment (jonsimantov) [reply .github/workflows/checks_secure.yml]: No longer necessary - it was necessary before I moved this into the checks_secure yml.

File: scripts/gha/dismiss_reviews.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  python scripts/gha/it_workflow.py --stage <stage> \
```
Comment (sunmou99) [inline scripts/gha/dismiss_reviews.py:18]: `python scripts/gha/dismiss_reviews.py`

Comment (jonsimantov) [reply scripts/gha/dismiss_reviews.py:18]: Eep :)

Comment (jonsimantov) [reply scripts/gha/dismiss_reviews.py:18]: BTW the github.py API is really nice. I'm going to move the other scripts I wrote over to use it soon.

File: scripts/gha/github.py [hunk 6]
```diff
@@ -155,9 +155,40 @@ def list_artifacts(token, run_id):
...
     with open(output_path, 'wb') as file:
         shutil.copyfileobj(response.raw, file)
+
+
+def dismiss_review(token, pull_number, review_id, message):
+  """https://docs.github.com/en/rest/reference/pulls#dismiss-a-review-for-a-pull-request"""
+  url = f'{FIREBASE_URL}/pulls/{pull_number}/reviews/{review_id}/dismissals'
+  headers = {'Accept': 'application/vnd.github.v3+json', 'Authorization': f'token {token}'}
+  data = {'message': message}
+  with requests.put(url, headers=headers, data=json.dumps(data),
```
Comment (sunmou99) [inline scripts/gha/github.py:171]: I believe is safe to use `requests_retry_session()` both here and for `get_reviews `.

**Review by sunmou99 — APPROVED** [review]
LGTM.
But I feel is good to have 2 demo PRs (1 for admin, the other for external user) to test it out.

**Comment by jonsimantov** [comment]
> LGTM.
> But I feel is good to have 2 demo PRs (1 for admin, the other for external user) to test it out.

Well it auto-merged, so we'll see if it breaks. :)

File: scripts/gha/github.py (hunk 6 above)
Comment (jonsimantov) [reply scripts/gha/github.py:171]: I'll add this in a follow up. If we have to retry this, should it just get the current page again, or should it get all the pages from scratch?

Comment (sunmou99) [reply scripts/gha/github.py:171]: For `get_reviews`, it's get current page again.

**Pull request merged**